packages/llamaindex/src/cloud/LLamaCloudFileService.ts

```diff
@@ -0,0 +1,171 @@
+import type { Metadata, NodeWithScore } from "@llamaindex/core/schema";
+import type { CloudConstructorParams } from "./constants.js";
+import { initService } from "./utils.js";
+
+const OUTPUT_DIR = "output/llamacloud";
+const FILE_DELIMITER = "$";
+
+export type LLamaCloudFileServiceConfigs = {
+  outputDir: string;
+  fileDelimiter: string;
+};
+
+export class LLamaCloudFileService {
+  params?: CloudConstructorParams;
+  configs: LLamaCloudFileServiceConfigs;
+
+  constructor(
+    params?: CloudConstructorParams & Partial<LLamaCloudFileServiceConfigs>,
+  ) {
+    this.configs = {
+      outputDir: params?.outputDir ?? OUTPUT_DIR,
+      fileDelimiter: params?.fileDelimiter ?? FILE_DELIMITER,
+    };
+    if (params) this.params = params;
+    initService(params);
+  }
+
+  /**
+   * get list of projects, each project contains a list of pipelines
+   */
+  public async getAllProjectsWithPipelines() {
+    try {
+      const projects = await ProjectsService.listProjectsApiV1ProjectsGet();
+      const pipelines =
+        await PipelinesService.searchPipelinesApiV1PipelinesGet();
+      return projects.map((project) => ({
+        ...project,
+        pipelines: pipelines.filter((p) => p.project_id === project.id),
+      }));
+    } catch (error) {
+      console.error("Error listing projects and pipelines:", error);
+      return [];
+    }
+  }
+
+  /**
+   * Upload a file to a pipeline in LlamaCloud
+   */
+  public async addFileToPipeline(
+    projectId: string,
+    pipelineId: string,
+    uploadFile: File | Blob,
+    customMetadata: Record<string, any> = {},
+  ) {
+    const file = await FilesService.uploadFileApiV1FilesPost({
+      projectId,
+      formData: {
+        upload_file: uploadFile,
+      },
+    });
+    const files = [
+      {
+        file_id: file.id,
+        custom_metadata: { file_id: file.id, ...customMetadata },
+      },
+    ];
+    await PipelinesService.addFilesToPipelineApiV1PipelinesPipelineIdFilesPut({
+      pipelineId,
+      requestBody: files,
+    });
+
+    // Wait 2s for the file to be processed
+    const maxAttempts = 20;
+    let attempt = 0;
+    while (attempt < maxAttempts) {
+      const result =
+        await PipelinesService.getPipelineFileStatusApiV1PipelinesPipelineIdFilesFileIdStatusGet(
+          {
+            pipelineId,
+            fileId: file.id,
+          },
+        );
+      if (result.status === "ERROR") {
+        throw new Error(`File processing failed: ${JSON.stringify(result)}`);
+      }
+      if (result.status === "SUCCESS") {
+        // File is ingested - return the file id
+        return file.id;
+      }
+      attempt += 1;
+      await new Promise((resolve) => setTimeout(resolve, 100)); // Sleep for 100ms
+    }
+    throw new Error(
+      `File processing did not complete after ${maxAttempts} attempts.`,
+    );
+  }
+
+  /**
+   * Get download URLs for a list of nodes
+   */
+  public async getDownloadFileUrls(nodes: NodeWithScore<Metadata>[]) {
```

Comment:
move this to CL - concern of this class should be just getting out and getting in files from llamacloud